presto-testing-docker/src/main/java/io/prestosql/testing/TemporaryDockerContainer.java

```diff
@@ -0,0 +1,215 @@
+import static java.lang.String.format;
+import static java.time.temporal.ChronoUnit.MINUTES;
+import static java.time.temporal.ChronoUnit.SECONDS;
+import static java.util.Objects.requireNonNull;
+
+public class TemporaryDockerContainer
+        implements Closeable
+{
+    private static final Logger LOG = Logger.get(TemporaryDockerContainer.class);
+
+    private final String hostIp = "127.0.0.1";
+    private final String image;
+    private final String hostName;
+    private final Map<String, String> environment;
+    private DockerClient dockerClient;
+    private String containerId;
+    private Map<Integer, Integer> hostPorts;
+
+    public TemporaryDockerContainer(String image, String hostName, List<Integer> ports, Map<String, String> environment)
+    {
+        this.image = requireNonNull(image, "image is null");
+        this.hostName = requireNonNull(hostName, "hostName is null");
+        this.environment = ImmutableMap.copyOf(requireNonNull(environment, "environment is null"));
+        try {
+            setUp(ports);
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> TemporaryDockerContainer.this.close()));
+    }
+
+    public void wait(CheckedRunnable verifier)
+    {
+        RetryPolicy<Object> retryPolicy = new RetryPolicy<>()
+                .withMaxDuration(Duration.of(5, MINUTES))
+                .withMaxAttempts(Integer.MAX_VALUE) // limited by MaxDuration
+                .onRetry(event -> LOG.info(format("Waiting for %s...", hostName)))
+                .withDelay(Duration.of(5, SECONDS));
+        Failsafe.with(retryPolicy).run(verifier);
+    }
+
+    private void setUp(List<Integer> ports)
+            throws Exception
+    {
+        dockerClient = DefaultDockerClient.fromEnv().build();
+        if (!image.endsWith("-SNAPSHOT") && dockerClient.listImages(DockerClient.ListImagesParam.byName(image)).isEmpty()) {
+            dockerClient.pull(image);
+        }
+        if (isDebug()) {
+            dockerClient.listContainers().stream()
+                    .filter(container -> container.image().equals(image))
+                    .forEach(container -> {
+                        checkState(containerId == null, "There are more than single container running");
+                        this.containerId = container.id();
+                    });
+            if (containerId != null) {
+                LOG.info("%s container already exists with id: %s", hostName, containerId);
+            }
+            else {
+                createContainer(ports);
+            }
+            if (hostPorts == null) {
+                calculateHostPorts(ports);
+            }
+        }
+        else {
+            dockerClient.listContainers().stream()
+                    .filter(container -> container.image().equals(image))
+                    .map(Container::id)
+                    .forEach(this::removeContainer);
```

Comment:
You cannot do this.
- i might have eg sqlserver container which i use for manual testing of the functionality i develop. I don't want it to be _implicitly_ killed when i run the tests. It would be a disaster.
- killing other containers will prevent us from using multiple containers in parallel (either to isolate tests, or to create specialized environment or configuration for a test)